rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,653 @@
+  // The N most recent authentication status records for this bot instance.
+  repeated BotInstanceStatusAuthentication latest_authentications = 4;
+  // The initial heartbeat status for this bot instance.
+  BotInstanceStatusHeartbeat initial_heartbeat = 5;
+  // The N most recent heartbeats for this bot instance.
+  repeated BotInstanceStatusHeartbeat latest_heartbeats = 6;
+}
+```
+
+The name used for a BotInstance will be a concatenation of the Bot name and its
+unique identifier (UUID).
+
+When storing the BotInstance in the backend, the key will be:
+`bot_instances/{bot_name}/{uuid}`. This will allow for efficient listing
+of BotInstances for a given Bot.
+
+Like agent heartbeats, the BotInstance will expire after a period of inactivity.
+This avoids the accumulation of ephemeral BotInstances.
+
+#### Recording Authentication Data
+
+Upon each join and renewal, the BotInstance record will be updated with an
+additional entry in the `status.authentications` field. If there is X entries,
+then the oldest entry will be removed. This prevents growth without bounds but
+also ensures that the original record is retained.
+
+In addition, the TTL of the BotInstance resource will be extended to cover the
+validity period of the issued certificate, plus a short additional time to allow
+for some imprecision.
+
+If a BotInstance does not exist, then one will be created. In the case that
+this occurs for a bot using the `token` join method and this is a renewal,
+a warning will be emitted and the initial generation of the BotInstance will
+be sourced from the certificates current generation counter. This behaviour
+will support the migration of existing `tbot` instances to the new BotInstance
+behaviour.
+
+#### Recording Heartbeat Data
+
+A new RPC will be added for submitting heartbeat data:
+
+```protobuf
+syntax = "proto3";
+
+package teleport.machineid.v1;
+
+service BotInstanceService {
+  // SubmitHeartbeat submits a heartbeat for a BotInstance.
+  rpc SubmitHeartbeat(SubmitHeartbeatRequest) returns (SubmitHeartbeatResponse);
+}
+
+// The request for SubmitHeartbeat.
+message SubmitHeartbeatRequest {
+  // The heartbeat data to submit.
+  BotInstanceStatusHeartbeat heartbeat = 1;
+}
+
+// The response for SubmitHeartbeat.
+message SubmitHeartbeatResponse {
+  // Empty
+}
+```
+
+The endpoint will have a special auth/authz check. RBAC will not be used and
+instead the endpoint will ensure that the presented client certificate is for
+the Bot linked to the instance.
+
+This endpoint will be called by `tbot` immediately after it has initially
+authenticated. After a heartbeat has successfully completed, another should be
+scheduled for a half hour after. A small amount of jitter should be added to
+the heartbeat period to avoid a thundering herd of heartbeats.
+
+If the heartbeat fails, then `tbot` should retry on an exponential backoff.
+
+##### Alternative: Submit Heartbeat Data on Join/Renew
```

Comment:
Before merging this, it would be helpful if each `Alternative` that we present but ultimately do not select lists why it was not chosen.

Reply:
It's been long enough now that I don't remember the order of events, but we now have a summary of why alternatives were not selected for (at least) public key fingerprints and the heartbeat RPC.